crates/wasmtime/src/ref.rs

```diff
@@ -2,224 +2,179 @@
-                Rc::ptr_eq(a, b)
-            }
-            _ => false,
-        }
+        self.inner == other.inner
     }
 
-    /// Returns a mutable reference to the host information if available.
-    /// # Panics
-    /// Panics if `ExternRef` is already borrowed or `ExternRef` is `Null`.
-    pub fn host_info(&self) -> Option<cell::RefMut<Box<dyn Any>>> {
-        match self {
-            ExternRef::Null => panic!("null"),
-            ExternRef::Ref(r) => r.0.host_info(),
-            ExternRef::Other(r) => {
-                let info = cell::RefMut::map(r.0.borrow_mut(), |b| &mut b.host_info);
-                if info.is_none() {
-                    return None;
-                }
-                Some(cell::RefMut::map(info, |info| info.as_mut().unwrap()))
-            }
-        }
+    /// Returns the host information for this `externref`, if previously created
+    /// with `set_host_info`.
+    pub fn host_info(&self) -> Option<Rc<RefCell<dyn Any>>> {
+        let store = crate::Store::upgrade(&self.store)?;
+        store.host_info(self)
     }
 
-    /// Sets the host information for an `ExternRef`.
-    /// # Panics
-    /// Panics if `ExternRef` is already borrowed or `ExternRef` is `Null`.
-    pub fn set_host_info(&self, info: Option<Box<dyn Any>>) {
-        match self {
-            ExternRef::Null => panic!("null"),
-            ExternRef::Ref(r) => r.0.set_host_info(info),
-            ExternRef::Other(r) => {
-                r.0.borrow_mut().host_info = info;
-            }
-        }
+    /// Set the host information for this `externref`, returning the old host
+    /// information if it was previously set.
+    pub fn set_host_info<T>(&self, info: T) -> Option<Rc<RefCell<dyn Any>>>
+    where
+        T: 'static + Any,
+    {
+        let store = crate::Store::upgrade(&self.store)?;
+        store.set_host_info(self, Some(Rc::new(RefCell::new(info))))
     }
-}
 
-impl fmt::Debug for ExternRef {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        match self {
-            ExternRef::Null => write!(f, "null"),
-            ExternRef::Ref(_) => write!(f, "externref"),
-            ExternRef::Other(_) => write!(f, "other ref"),
-        }
+    /// Remove the host information for this `externref`, returning the old host
+    /// information if it was previously set.
+    pub fn remove_host_info(&self) -> Option<Rc<RefCell<dyn Any>>> {
+        let store = crate::Store::upgrade(&self.store)?;
+        store.set_host_info(self, None)
     }
 }
 
-struct ContentBox<T> {
-    content: T,
-    host_info: Option<Box<dyn Any>>,
-    externref_data: Weak<dyn InternalRefBase>,
+impl fmt::Debug for ExternRef {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let ExternRef { inner, store: _ } = self;
+        f.debug_struct("ExternRef")
+            .field("inner", &inner)
+            .field("store", &"..")
```

Comment:
Could this pass in `self.store()` and defer to `Debug for Store`?

Reply:
`Store` doesn't implement `Debug`, but I can add an impl.